src/ucp/proto/proto_select.c

```diff
@@ -339,41 +291,23 @@ static ucs_status_t ucp_proto_select_elem_add_envelope(
             thresh_elem = ucs_array_append(thresholds,
                                            return UCS_ERR_NO_MEMORY);
 
+            ucs_assert(proto_idx < UINT16_MAX);
             thresh_elem->max_msg_length  = envelope_elem->max_length;
             proto_config                 = &thresh_elem->proto_config;
             proto_config->proto          = ucp_protocols[proto->proto_id];
             proto_config->priv           = proto_priv;
             proto_config->ep_cfg_index   = ep_cfg_index;
             proto_config->rkey_cfg_index = rkey_cfg_index;
             proto_config->select_param   = *select_param;
+            proto_config->init_elem      = proto;
             *last_proto_idx              = proto_idx;
         }
 
-        /* Do not unite performance ranges, since they could have same final
-         * result but calculated differently. And we want to track the whole
-         * calculation process.
-         */
-        range = ucs_array_append(perf_ranges, return UCS_ERR_NO_MEMORY);
-
-        caps_range = ucp_proto_caps_range_find(&proto->caps, range_start);
-        ucs_assert(caps_range != NULL); /* for cppcheck */
-
-        range->max_length = envelope_elem->max_length;
-        ucp_proto_perf_copy(range->perf, caps_range->perf);
-
-        ucs_memunits_range_str(range_start, envelope_elem->max_length,
-                               range_str, sizeof(range_str));
-
-        range->node = ucp_proto_perf_node_new_select(
-                node_name, envelope_elem->index, "%s %s",
-                ucp_proto_id_field(proto->proto_id, name), range_str);
-
-        /* Add all candidates as children */
-        UCS_DYNAMIC_BITMAP_FOR_EACH_BIT(child_proto_idx, proto_mask) {
-            proto       = &ucs_array_elem(&proto_init->protocols,
-                                          child_proto_idx);
-            child_range = ucp_proto_caps_range_find(&proto->caps, range_start);
-            ucp_proto_perf_node_add_child(range->node, child_range->node);
+        /* Print detailed protocol selection data to a user-configured path */
+        if (!internal && !ucs_string_is_empty(proto_info_dir)) {
+            ucp_proto_select_write_info(worker, proto_init, proto_mask,
+                                        proto_idx, &thresh_elem->proto_config,
+                                        range_start, envelope_elem->max_length);
```

Comment:
That was changed because in the master we create select perf node and store it in structure array called `ucp_proto_ranges_t` which was a part of `ucp_proto_select_elem_t`. Now there is no `ucp_proto_perf_range_t` structure since there is no `ucp_proto_caps_t` structure and therefore we do not have place to store selection perf nodes together with `max_length` that allows us to build performance graph. 

That's why I decided to build selection perf node on flight during `_add_envelope` function, draw the perf graph for some certain range and then deref this selection perf node. Since for that I need to obtain `proto_mask` for certain ranges I decided to move it here.

We still can create selection perf nodes in that place, store them together with meta information in some container (like it was done with `ucp_proto_ranges_t` and propagate all this information to `ucp_proto_select_elem_trace` but I am not sure that this worth it.

Reply:
Added TODO